### ISSUE TITLE
Ability to set a delay between retries and specifying on which exceptions it retries

### ISSUE THREAD

**Issue by pandoras-toolbox** [gradle/test-retry-gradle-plugin#74]
It would be good to be able to define a delay between retries because it sometimes can make much more sense to retry a test after lets say 30 seconds. If tests are executed in parallel then this delay should not be too blocking.

Also it would be nice to be able to specify on which exceptions and exception messages to retry.

An example for these things is Failsafe with a relatively configurable retry policy.

**Comment by ldaley** [comment]
@pandoras-toolbox can you please provide examples of why you would use this functionality? That is, in what kind of cases is this needed?

**Comment by pandoras-toolbox** [comment]
In a few cases, not many I admit, it can make a difference I thought. When nightly run end to end tests are run and something is being deployed to an external system while the test is running, then a test might fail. When the test is rerun immediately the chances are high that it fails again. But if it would wait for lets say three minutes before repeating, then the chances can be higher that it succeeds because the deployment has finished until then (when there is no blue/green deployment). Or if a microservice process of the tested application has died, it might be detected automatically in certain intervals by a monitoring tool and be restarted automatically, making it more likely that the test will pass when rerun. A third reason for it might be that the test failed because of an unusual high load at that time. When run a few minutes later that load is more likely to have changed than if rerun immediately.

What maybe would be nice is to have a random delay between a minimum and maximum duration, like something between 1 and 2 minutes. When tests are run in parallel and many fail because of the same reason, then this random duration can reduce the number of retries needed and provide more information about the behavior of the system.

What came into my mind writing this is that for certain tests a retry might not make sense because the test does not involve external systems which are not in control for instance. Whether a test is restarted or not and with which delay, it would maybe be good to be able to configure this per test.

Last but not least, if a test is rerun can be dependent on the test failure itself, for instance the type of exception or the exception message. For instance, in a test the programmer can throw a certain exception, and that could mean that the test can be restarted or that it should not be restarted.

Actually that is all possible with Failsafe dependency, but with the disadvantage that the test report does not give any information about the retries. With your plugin I can see in Allure report tool the number of retries and their results.

**Comment by ldaley** [comment]
@pandoras-toolbox 

> Actually that is all possible with Failsafe dependency

Can you please point me at some documentation for that. I haven't been able to find anything.

**Comment by pandoras-toolbox** [comment]
This is what I meant: https://github.com/jhalterman/failsafe

**Comment by ldaley** [comment]
@pandoras-toolbox I would like to avoid going down this path, at least for now. 

It's my assertion that such fine grained control, and resilience against such systemically volatile infrastructure, should be part of the test logic itself. 

I'm going to close this off for now as `wontfix`, but am open to more discussion. We can always reopen the issue if a compelling rationale emerges.

**Comment by pandoras-toolbox** [comment]
Well, then what sense does it make to retry at all a millisecond later? The situation is very likely still the same and so the outcome.
